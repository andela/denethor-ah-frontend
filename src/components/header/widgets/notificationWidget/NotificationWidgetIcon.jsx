import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBell } from '@fortawesome/free-regular-svg-icons';

const NotificationWidgetIcon = ({ toggleBubble, notifications }) => (
  <div className="header__notification-widget__icon">
    <button onClick={toggleBubble}>
      <FontAwesomeIcon
        className={
          `header__notification-widget__icon__bell${notifications.some(({ read }) => !read)
            ? ' unread'
            : ''}`}
        icon={faBell}
        size="2x"
        onClick={toggleBubble}
      />
    </button>
  </div>
);

NotificationWidgetIcon.propTypes = {
  toggleBubble: PropTypes.func,
  notifications: PropTypes.array
};

const mapStateToProps = ({ profile: { imageUrl } }) => ({ imageUrl });

export default connect(mapStateToProps)(NotificationWidgetIcon);
