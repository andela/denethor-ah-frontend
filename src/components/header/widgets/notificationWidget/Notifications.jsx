import React from 'react';
import uuid from 'uuid/v1';
import NotificationItem from './Notification';

const Notifications = (notifications) => () => (
  <div className="header__notification-widget__preview-pane">
    <ul>
    {
      notifications[0]
        ?
          notifications.map((notification) =>
            <NotificationItem key={uuid()} notification={notification} />
          )
        : 
          <li className="header__notification-widget__preview-pane__no-notifications">
            <p>No notifications</p>
          </li>
      }
    </ul>
  </div>
);

export default Notifications;
