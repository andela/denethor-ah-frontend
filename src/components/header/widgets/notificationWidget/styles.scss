.header__notification-widget {
  display: flex;
  align-items: center;
  .header__notification-widget__icon {
    button {
      background-color: transparent;
      border: none;
      outline: none;
      .header__notification-widget__icon__bell {
        color: #EBB03D;
        &.unread {
          animation: ring 4s .7s ease-in-out infinite, glowing 2s infinite,;
          transform-origin: 50% 4px;
        }
      }
    }
  }
  ul {
    li {
      a {
        display: flex;
        width: 100%;
        p {
          width: 100%;
          &.unread {
            font-family: Montserrat-SemiBold;
          }
        }
        small {
          color: #777;
          padding-left: 0.5rem;
        }
      }
    }
  }

  .header__widget__bubble__content {
    &:after {
      right: 4.7rem;
    }
  }
}

.notification-widget__toast {
  color: #333;
  position: fixed;
  top: 5rem;
  a {
    text-decoration: none;
    color: #333;
  }
}

@keyframes glowing {
  0% {
    color: darken(#EBB03D, 10%);
    filter: drop-shadow(0 0 -10px #fff); }
  40% {
    color: lighten(#EBB03D, 10%);
    filter: drop-shadow(0 0 20px lighten(#EBB03D, 20%));
  }
  60% {
    color: lighten(#EBB03D, 10%);
    filter: drop-shadow(0 0 20px lighten(#EBB03D, 20%));
  }
  100% {
    color: darken(#EBB03D, 10%);
    filter: drop-shadow(0 0 -10px #fff);
  }
}

@keyframes ring {
  0% { transform: rotate(0); }
  1% { transform: rotate(30deg); }
  3% { transform: rotate(-28deg); }
  5% { transform: rotate(34deg); }
  7% { transform: rotate(-32deg); }
  9% { transform: rotate(30deg); }
  11% { transform: rotate(-28deg); }
  13% { transform: rotate(26deg); }
  15% { transform: rotate(-24deg); }
  17% { transform: rotate(22deg); }
  19% { transform: rotate(-20deg); }
  21% { transform: rotate(18deg); }
  23% { transform: rotate(-16deg); }
  25% { transform: rotate(14deg); }
  27% { transform: rotate(-12deg); }
  29% { transform: rotate(10deg); }
  31% { transform: rotate(-8deg); }
  33% { transform: rotate(6deg); }
  35% { transform: rotate(-4deg); }
  37% { transform: rotate(2deg); }
  39% { transform: rotate(-1deg); }
  41% { transform: rotate(1deg); }

  43% { transform: rotate(0); }
  100% { transform: rotate(0); }
}
