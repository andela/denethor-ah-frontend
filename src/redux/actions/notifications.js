import socket from '../../utils/socket';
import { REMOVE_NOTIFICATION } from './types';

export const removeNotifications = (articleId) => async (dispatch, getState) => {
  dispatch({
    type: REMOVE_NOTIFICATION,
    payload: articleId
  });

  const { notifications } = getState();

  socket.emit('setRemainingNotifications', notifications);
}
