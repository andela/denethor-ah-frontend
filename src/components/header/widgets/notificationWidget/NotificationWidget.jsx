import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter, Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import Bubble from '../bubble';
import NotificationWidgetIcon from './NotificationWidgetIcon';
import Notifications from './Notifications';
import removeNotifications from '../../../../redux/actions/notifications';
import './styles.scss';


export class NotificationWidget extends Component {
  state = {
    showBubble: false,
    newNotifications: 0,
    toastId: 'notification'
  }

  componentDidMount () {
    const {
      props: { notifications = [], removeNotifications },
      state: { toastId }
    } = this;
  
    const newNotification = notifications.find(({ read }) => !read);
    const newNotificationsInProps = notifications.reduce((count, { read }) => count + (read ? 0 : 1), 0);

    if(newNotification) {
      const { message, articleId } = newNotification;

      if (!toast.isActive(toastId)) {
        const Notification = ({ closeToast }) => (
          <Link
          to={`/articles/${articleId}`}
          onClick={() => {
            removeNotifications(articleId);
            closeToast()
          }}>
            {message}
          </Link>
        );

        toast(<Notification />, {
          toastId,
          autoClose: 10000,
          hideProgressBar: true,
          className: 'notification-widget__toast',
          closeOnClick: false
        });
      }
      const newNotifications = newNotificationsInProps;

      this.setState({ newNotifications });
    }
  }

  componentDidUpdate () {
    const {
      props: { notifications = [], removeNotifications },
      state: { newNotifications: newNotificationsInState, toastId }
    } = this;

    const newNotificationsInProps = notifications.reduce((count, { read }) => count + (read ? 0 : 1), 0);

    if(newNotificationsInProps > newNotificationsInState) {
      const { message, articleId } = notifications.find(({ read }) => !read);

      if (!toast.isActive(toastId)) {
        const Notification = ({ closeToast }) => (
          <Link
          to={`/articles/${articleId}`}
          onClick={() => {
            removeNotifications(articleId);
            closeToast()
          }}>
            {message}
          </Link>
        );

        toast(<Notification />, {
          toastId,
          autoClose: 10000,
          hideProgressBar: true,
          className: 'notification-widget__toast',
          closeOnClick: false
        });
      }
      const newNotifications = newNotificationsInProps;

      this.setState({ newNotifications });
    }
  }

  toggleBubble = (showBubble) => {
    if(showBubble !== undefined) {
      return this.setState({ showBubble });
    }
    this.setState(({ showBubble }) => ({ showBubble: !showBubble }));
  }

  render() {
    const {
      toggleBubble,
      props: { notifications = [] },
      state: { showBubble }
    } = this;

    return (
      <div className="header__notification-widget">
        <NotificationWidgetIcon notifications={notifications} toggleBubble={toggleBubble} />
        {showBubble && <Bubble BubbleContent={Notifications(notifications)} toggleBubble={toggleBubble}/>}
      </div>
    );
  }
}

NotificationWidget.propTypes = {
  history: PropTypes.object,
  notifications: PropTypes.array,
  removeNotifications: PropTypes.func
}

const mapStateToProps = ({ notifications }) => ({ notifications });

const mapDispatchToProps = (dispatch) => ({
  removeNotifications: (articleId) => dispatch(removeNotifications(articleId)),
});

export default (connect(mapStateToProps, mapDispatchToProps)(withRouter(NotificationWidget)));
