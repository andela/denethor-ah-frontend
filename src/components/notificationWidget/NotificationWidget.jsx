import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import uuid from 'uuid/v1';
import PropTypes from 'prop-types';
import NotificationsPreviewItem from './NotificationsPreviewItem';
import RoundedImage from '../RoundedImage';
import './styles.scss';


export class NotificationWidget extends Component {
  state = {
    showPreview: false,
    showBadge: false,
  }

  componentDidMount = () => {
    this.setState({
      showPreview: false,
      pathname: this.props.history.location.pathname
    });
  }

  componentDidUpdate = () => {
    const {
      props: { notifications, history: { location: { pathname } } },
      state: { pathname: statePathname, showPreview }
    } = this;
    if(notifications[0] && !this.state.showBadge) {
      this.setState({ showBadge: true });
    } else if (!notifications[0] && this.state.showBadge) {
      this.setState({ showBadge: false });
    }

    if((pathname !== statePathname) && showPreview) {
      this.setState({
        showPreview: false,
        pathname
      });
    }
  }

  handleToggle = () => {
    const {
      props: { history: { location: { pathname } } },
      state: { pathname: statePathname }
    } = this;

    const widget = document.querySelector('.notifications-widget__preview-pane--wrapper');

    if(pathname !== statePathname) {
      this.setState({ pathname });
    }

    this.setState({ showPreview: true });

    window.onclick = (event) => {
      if (event.target == widget) {
        this.setState({ showPreview: false });
      }
    }
}

  render() {
    const { imageUrl, notifications, id, handleLogout } = this.props;
    return (
      <div
        className="notifications-widget"
      >
        <button
          className="notifications-widget__rounded-image"
          onClick={this.handleToggle}
        >
          <div className={this.state.showBadge ? 'notifications-widget__rounded-image__badge' : 'hidden'}></div>
          <RoundedImage imageSource={imageUrl || '/assets/img/placeholder-profile-picture.png'} alt="notifications widget" />
        </button>
        <div className={
          `notifications-widget__preview-pane--wrapper${ this.state.showPreview ? '' : ' hidden'}`
        }>
          <div className={
            `notifications-widget__preview-pane${ this.state.showPreview ? '' : ' hidden'}`
          }>
            <ul>
            {
              notifications[0]
                ?
                  notifications.map((notification) =>
                    <NotificationsPreviewItem key={uuid()} notification={notification} ownerId={id} />
                  )
                : 
                  <li className="notifications-widget__preview-pane--no-notifications">
                    <p>No notifications</p>
                  </li>
              }
              <li><button onClick={handleLogout}>Log Out</button></li>
            </ul>
          </div>
        </div>
      </div>
    );
  }
}

NotificationWidget.propTypes = {
  id: PropTypes.string,
  handleLogout: PropTypes.func,
  history: PropTypes.object,
  imageUrl: PropTypes.string,
  notifications: PropTypes.array,
  getNotifications: PropTypes.func
}

const mapStateToProps = ({ profile: { id, imageUrl }, notifications }) => ({ id, imageUrl, notifications });

export default withRouter(connect(mapStateToProps)(NotificationWidget));
