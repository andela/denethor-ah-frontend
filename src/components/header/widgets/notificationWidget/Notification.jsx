import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { removeNotifications } from '../../../../redux/actions/notifications';

moment.updateLocale('en', {
  relativeTime : {
      future: "in %s",
      past:   "%s",
      s  : '%ds',
      ss : '%ds',
      m:  "%dm",
      mm: "%dm",
      h:  "%dh",
      hh: "%dh",
      d:  "%dd",
      dd: "%dd",
      M:  "a month",
      MM: "%d months",
      y:  "a year",
      yy: "%d years"
  }
});


export const Notification = props => {
  const { message, articleId, time, read } = props.notification;
  
  return (
    <li>
      <Link to={`/articles/${articleId}`} onClick={() => props.removeNotifications(articleId)}>
        <p className={read ? '' : 'unread'}>{message}</p>
        <span><small>{moment(time).fromNow()}</small></span>
      </Link>
    </li>
  );
}

Notification.propTypes = {
  notification: PropTypes.object,
  removeNotifications: PropTypes.func
}

export const mapDispatchToProps = (dispatch) => ({
  removeNotifications: (articleId) => dispatch(removeNotifications(articleId))
});

export default connect(undefined, mapDispatchToProps)(Notification);
