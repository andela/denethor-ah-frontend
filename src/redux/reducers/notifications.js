import { SET_NOTIFICATIONS, REMOVE_NOTIFICATION } from '../actions/types';

const notificationsDefaultState = [];

const notificationsReducer = (state = notificationsDefaultState, action) => {
  switch (action.type) {
    case SET_NOTIFICATIONS:
      return action.payload;

    case REMOVE_NOTIFICATION:
      return state.map((notification) => ({
        ...notification,
        ...(notification.articleId === action.payload && { read: true })
      }));

    default:
      return state;
  }
};

export default notificationsReducer;
