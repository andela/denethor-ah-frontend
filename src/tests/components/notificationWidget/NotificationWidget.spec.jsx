import React from 'react';
import uuid from 'uuid/v1';
import { shallow } from 'enzyme';
import { NotificationWidget } from '../../../components/header/widgets/notificationWidget/NotificationWidget';
import notifications from './mock-notifications';

jest.mock('../../../utils/socket');
jest.mock('uuid/v1');

test('Should render static component', () => {
  uuid.mockReturnValue(expect.any(String));

  const props = {
    notifications,
    id: 'id',
    history: {
      location: {
        pathname: '/'
      }
    }
  }
  const wrapper = shallow(<NotificationWidget { ...props } />);

  expect(wrapper).toMatchSnapshot();
});
