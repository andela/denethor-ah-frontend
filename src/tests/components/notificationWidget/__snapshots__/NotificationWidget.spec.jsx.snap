// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render static component 1`] = `
<div
  className="header__notification-widget"
>
  <Connect(NotificationWidgetIcon)
    notifications={
      Array [
        Object {
          "articleId": "1",
          "message": "this is the first notification",
          "time": "8675678uygi",
        },
        Object {
          "articleId": "2",
          "message": "this is the second notification",
          "time": "8675678uygisd3",
        },
        Object {
          "articleId": "3",
          "message": "this is the third notification",
          "read": true,
          "time": "8675678uygjkhi",
        },
      ]
    }
    toggleBubble={[Function]}
  />
</div>
`;
